src/data/sequence.rs

```diff
@@ -0,0 +1,788 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+// #![allow(dead_code)] // for the draft PR only
+
+use crate::authorization::access_control::{
+    AccessListTrait, AccessType, PrivateAccessList, PrivateUserAccess, PublicAccessList,
+    PublicUserAccess,
+};
+use crate::shared_data::{
+    to_absolute_range, to_absolute_version, Address, ExpectedVersions, Kind, NonSentried, Owner,
+    Sentried, User, Value, Version,
+};
+use crate::{Error, PublicKey, Result, XorName};
+use serde::{Deserialize, Serialize};
+use std::fmt::{self, Debug, Formatter};
+
+pub type PublicSentriedSequence = SequenceBase<PublicAccessList, Sentried>;
+pub type PublicSequence = SequenceBase<PublicAccessList, NonSentried>;
+pub type PrivateSentriedSequence = SequenceBase<PrivateAccessList, Sentried>;
+pub type PrivateSequence = SequenceBase<PrivateAccessList, NonSentried>;
+pub type Values = Vec<Value>;
+
+/// A representation of data at some version.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Default, Debug)]
+pub struct DataEntry {
+    /// Data version
+    pub version: u64,
+    /// Data value
+    pub value: Vec<u8>,
+}
+
+impl DataEntry {
+    /// Returns a new instance of a data entry.
+    pub fn new(version: u64, value: Vec<u8>) -> Self {
+        Self { version, value }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash)]
+pub struct SequenceBase<P, S> {
+    address: Address,
+    data: Values,
+    access_list: Vec<P>,
+    // This is the history of owners, with each entry representing an owner.  Each single owner
+    // could represent an individual user, or a group of users, depending on the `PublicKey` type.
+    owners: Vec<Owner>,
+    _flavour: S,
+}
+
+/// Common methods for all `Sequence` flavours.
+impl<C, S> SequenceBase<C, S>
+where
+    C: AccessListTrait,
+    S: Copy,
+{
+    /// Returns true if the provided access type is allowed for the specific user (identified y their public key).
+    pub fn is_allowed(&self, user: PublicKey, access: AccessType) -> bool {
+        if let Some(owner) = self.owner_at(Version::FromEnd(1)) {
```

Comment:
this manual version value seems suspicious. 
at least a comment explaining why we can do so?

Reply:
Yeah I know and agree. (Additionally, it might actually be FromEnd(0) to get last entry of the vec..)
The aim is top get the `current entry`, which is the current owner, without specifying an index.
It would only work if there are no race conditions. Do you have suggestion for this?

Reply:
I don't have a better solution to the issue, however, I'd at least use a const and put some comment to explain why doing so.
so that, in case we do need to change from `FromEnd(1)` to `FromEnd(0)`, it will be more explicit.